src/hasSubObject.test.ts

```diff
@@ -126,47 +211,395 @@ describe("typing", () => {
-        expectTypeOf(obj).toMatchTypeOf<{ readonly a: "a"; b?: number }>();
+        expectTypeOf(obj).toMatchTypeOf<{ a: "a"; b?: number }>();
       } else {
         expectTypeOf(obj).toMatchTypeOf<{ a?: string; b?: number }>();
       }
     });
+
+    it("narrows with sub-object union type field", () => {
+      const obj: { a?: string; b?: number; c?: boolean } = {};
+
+      if (hasSubObject(obj, { c: true as boolean | number })) {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a?: string;
+          b?: number;
+          c: boolean;
+        }>();
+      } else {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a?: string;
+          b?: number;
+          c?: boolean;
+        }>();
+      }
+    });
+
+    it("narrows union types", () => {
+      const obj: { a?: string; b?: number; c?: number | string } = {};
+
+      if (hasSubObject(obj, { c: true } as { c?: boolean | number })) {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a?: string;
+          b?: number;
+          c?: number;
+        }>();
+      } else {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a?: string;
+          b?: number;
+          c?: number | string;
+        }>();
+      }
+    });
+
+    it("narrows nested fields with sub-object union types", () => {
+      const obj = {
+        a: { foo: "test", bar: true },
+        b: { foo: "test", bar: true },
+      };
+
+      if (
+        hasSubObject(obj, {
+          a: { foo: 12 as number | string, bar: true },
+          b: { foo: "test", bar: true as boolean | number },
+        })
+      ) {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a: { foo: string; bar: boolean };
+          b: { foo: string; bar: boolean };
+        }>();
+      } else {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a: { foo: string; bar: boolean };
+          b: { foo: string; bar: boolean };
+        }>();
+      }
+    });
+
+    it("narrows nested fields with union types", () => {
+      const obj = {
+        a: { foo: "test" as number | string, bar: true },
+        b: { foo: "test", bar: true as boolean | number },
+      };
+
+      if (
+        hasSubObject(obj, {
+          a: { foo: 12, bar: true },
+          b: { foo: "test", bar: true },
+        })
+      ) {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a: { foo: number; bar: boolean };
+          b: { foo: string; bar: boolean };
+        }>();
+      } else {
+        expectTypeOf(obj).toMatchTypeOf<{
+          a: { foo: number | string; bar: boolean };
+          b: { foo: string; bar: boolean | number };
+        }>();
+      }
+    });
   });
 
   describe("data-last", () => {
-    it("must have matching keys and values", () => {
+    it("returns boolean", () => {
       expectTypeOf(
         pipe({ a: 2 }, hasSubObject({ a: 1 })),
       ).toEqualTypeOf<boolean>();
+    });
+
+    it("doesn't require sub-object to have all keys", () => {
+      // ok
+      hasSubObject({ b: 2 })({ a: 1, b: 2 });
```

Comment:
**nit:** let's cut the data-last tests that aren't in a pipe—we don't usually test these

Reply:
@cjquines I think it's better to leave them because their behavior can be different. Tests with `pipe` depends on `pipe`'s implementation.

Reply:
i agree this is usually the case, but by far the most common use case for data-last functions are in pipes, and testing it there brings it closer to what people use

Reply:
@cjquines That's why I left both. One tests it as it is and another tests a more common scenario.